### DISCUSSION TITLE
How do I trace io_uring when tuning for performance?

### DISCUSSION THREAD

**Discussion by ghost** [axboe/liburing#1196] [General]
I am trying to find out the sweet spot for timeout and number of events to wait for the event loop that uses  io_uring_submit_and_wait_timeout and multishot recvs.

(For the record, even without timeout i.e. just using io_uring_submit_and_wait(1) netted great result but I feel like I'm leaving extra performance on table without making efforts to optimize batching further.)

What would be the smart way to setup experiments to optimize parameters other than just bruteforcing benchmarks?

Is there perf tracepoints that might be useful to observe?

**Comment by axboe** [comment]
There's a good amount of tracepoints, check debugfs:

```
root@m2max:~# ls /sys/kernel/debug/tracing/events/io_uring/
enable		       io_uring_fail_link	  io_uring_req_failed
filter		       io_uring_file_get	  io_uring_short_write
io_uring_complete      io_uring_link		  io_uring_submit_req
io_uring_cqe_overflow  io_uring_local_work_run	  io_uring_task_add
io_uring_cqring_wait   io_uring_poll_arm	  io_uring_task_work_run
io_uring_create        io_uring_queue_async_work
io_uring_defer	       io_uring_register
```

**Comment by koonwen** [comment]
@kafka64 I wrote a tracing tool for uring [here](https://github.com/koonwen/uring-trace). It constructs a trace file from above mentioned tracepoints to be loaded into perfetto to visualize how your requests are interfacing with uring. You may find it useful

![gif](https://github.com/koonwen/uring-trace/blob/main/assets/Recording.gif?raw=true)

**Reply by ghost** [reply]
This looks amazing. I will check it out :D
